sigstore/sign.py

```diff
@@ -211,22 +212,26 @@ def sign_dsse(
         )
 
         # Sign the statement, producing a DSSE envelope
-        content = dsse._sign(self._private_key, input_)
-
-        # Create the proposed DSSE log entry
-        proposed_entry = rekor_types.Dsse(
-            spec=rekor_types.dsse.DsseSchema(
-                # NOTE: mypy can't see that this kwarg is correct due to two interacting
-                # behaviors/bugs (one pydantic, one datamodel-codegen):
-                # See: <https://github.com/pydantic/pydantic/discussions/7418#discussioncomment-9024927>
-                # See: <https://github.com/koxudaxi/datamodel-code-generator/issues/1903>
-                proposed_content=rekor_types.dsse.ProposedContent(  # type: ignore[call-arg]
-                    envelope=content.to_json(),
-                    verifiers=[b64_cert.decode()],
+        content: dsse.Envelope = None
+        proposed_entry: rekor_types.ProposedEntry = None
+        if type(input_) is dsse.Statement:
+            content = dsse._sign(self._private_key, input_)
+            # Create the proposed DSSE log entry
+            proposed_entry = rekor_types.Dsse(
+                spec=rekor_types.dsse.DsseSchema(
+                    # NOTE: mypy can't see that this kwarg is correct due to two interacting
+                    # behaviors/bugs (one pydantic, one datamodel-codegen):
+                    # See: <https://github.com/pydantic/pydantic/discussions/7418#discussioncomment-9024927>
+                    # See: <https://github.com/koxudaxi/datamodel-code-generator/issues/1903>
+                    proposed_content=rekor_types.dsse.ProposedContent(  # type: ignore[call-arg]
+                        envelope=content.to_json(),
+                        verifiers=[b64_cert.decode()],
+                    ),
                 ),
-            ),
-        )
-
+            )
+        elif type(input_) is dsse.RawPayload:
+            content = dsse._sign_payload(self._private_key, input_)
+            # TODO: figure out an entry that works.
```

Comment:
NB: this is still a hard blocker: the client specification requires bundles to have a transparency log entry, so we can't just skip it when the signed-over input isn't a DSSE.

Reply:
I think there are two possible resolutions here, both of which require some kind of upstream change:

* Option 1: Rekor learns about non-JSON DSSE payloads. I don't know what the timeline for this would be 🙂 
* Option 2: The client spec clarifies/expresses a judgement about using `dsse` bundles with non-`dsse` Rekor entry types (such as `hashedrekord`, which would work fine here *if* the DSSE envelope is canonicalized).

Reply:
In case it helps: the model signing use case would contain JSON as payload. However, as things stand now rekor solely supports in-toto statements.